File: api/app/rate_limit.py
"""
Fixed-window in-process rate limiter.

Sufficient for a single API instance; swap the dict for Redis when you scale
horizontally — the call signature is deliberately identical.
"""

from __future__ import annotations

import time
from threading import Lock

from fastapi import HTTPException, status

_buckets: dict[str, tuple[int, float]] = {}
_lock = Lock()


def rate_limit(key: str, limit: int, window_seconds: int) -> None:
    """Raises 429 when the caller has exceeded `limit` hits in the window."""
    now = time.monotonic()

    with _lock:
        count, reset_at = _buckets.get(key, (0, 0.0))

        if reset_at <= now:
            _buckets[key] = (1, now + window_seconds)
            return

        if count >= limit:
            retry_after = max(1, int(reset_at - now))
            raise HTTPException(
                status.HTTP_429_TOO_MANY_REQUESTS,
                detail="Rate limit exceeded. Try again shortly.",
                headers={"Retry-After": str(retry_after)},
            )

        _buckets[key] = (count + 1, reset_at)

        # Opportunistic cleanup so the dict cannot grow without bound.
        if len(_buckets) > 10_000:
            for stale_key, (_, expiry) in list(_buckets.items()):
                if expiry <= now:
                    _buckets.pop(stale_key, None)

File: web/lib/rate-limit.ts
import 'server-only'

/**
 * Fixed-window in-memory rate limiter.
 *
 * Good enough for a single instance and for protecting the public free-tool
 * endpoints from casual abuse. Behind more than one instance, swap the map for
 * Redis — the call signature is deliberately identical to a Redis-backed one.
 */

interface Bucket {
  count: number
  resetAt: number
}

const globalBuckets = globalThis as unknown as { __oh_rate_buckets?: Map<string, Bucket> }
const buckets: Map<string, Bucket> = (globalBuckets.__oh_rate_buckets ??= new Map())

export interface RateLimitResult {
  ok: boolean
  remaining: number
  retryAfterSeconds: number
}

export function rateLimit(key: string, limit: number, windowSeconds: number): RateLimitResult {
  const now = Date.now()
  const bucket = buckets.get(key)

  if (!bucket || bucket.resetAt <= now) {
    buckets.set(key, { count: 1, resetAt: now + windowSeconds * 1000 })
    return { ok: true, remaining: limit - 1, retryAfterSeconds: 0 }
  }

  bucket.count += 1

  if (bucket.count > limit) {
    return { ok: false, remaining: 0, retryAfterSeconds: Math.ceil((bucket.resetAt - now) / 1000) }
  }

  // Opportunistic cleanup so the map cannot grow without bound.
  if (buckets.size > 5000) {
    for (const [k, v] of buckets) if (v.resetAt <= now) buckets.delete(k)
  }

  return { ok: true, remaining: limit - bucket.count, retryAfterSeconds: 0 }
}

/** Best-effort client IP from the usual proxy headers. */
export function clientIp(headers: Headers): string {
  return (
    headers.get('x-forwarded-for')?.split(',')[0]?.trim() ||
    headers.get('x-real-ip') ||
    headers.get('cf-connecting-ip') ||
    'unknown'
  )
}
